### ISSUE TITLE
Per-project dockerized Python virtual environment

### ISSUE THREAD

**Issue by sfavazza** [emacs-lsp/lsp-docker#76]
Hello there, as mentioned in the object, I would like to create a per-project docker, such that each of them contain a project-specific Python virtual environment where installing (among others) the pylsp package.

The idea would be to have some dir-locals explicitly selecting the dockerized version of the `pylsp` server instead of the system-wide installed one.

I read multiple times the documentation of this repo and the one from the lsp-mode, though I cannot yet achieve my goal.

In a first attempt I tried with the following `.lsp-docker.yml` file:
```yml
lsp:
  server:
    type: docker
    subtype: image
    name: "mylspserver" # manually created image with this name
    server: "pylsp" 
    launch_command: "pylsp"
  mappings:
    - source: "."
      destination: "/code"
```

Hence invoking `lsp-docker-register` would install this server, but I guess this simply replaces the global lang-server for Python, as switching to another project still shows the same server connection pointing to `mylspserver`.

I then tried to create a unique ID client in lsp with:
```elisp
(lsp-register-client
 (make-lsp-client :new-connection (lsp-stdio-connection "pylsp")
                  :activation-fn (lsp-activate-on "python")
                  :server-id 'my-fancy-pylsp))
```
Changing the `server` key-word in `.lsp-docker.yml` to `my-fancy-pylsp` allow me to keep `pylsp` server intact and just add the `my-fancy-pylsp`. Though I don't know how to prevent other projects from pointing at `my-fancy-pylsp`.

This seconds approach is not nice as it implies to always create a new server (for each new project) with a unique name that is not local to the project.

Am I missing something here? Is it possible at all to achieve my goal?

If a neat solution can be found, I am willing to add this use-case to the `lsp-docker` documentation with a MR.

**Comment by sfavazza** [comment]
In my current scenario they are independent one each other, no path
relation.

But I cannot exclude I will have a project under another one in the future.
A solution working for both scenario would be ideal, though for now let's
focus on the first one (no inter-project paths relations).

**Comment by factyy** [comment]
Ok, I'll take a look, but please keep in mind, that the first scenario you mentioned above is not about replacing the `pylsp` server globally, but registering a new one with a higher priority. And when you try to connect to the server the paths have to match (imagine you type `lsp` , then your current buffer file has to reside **inside** a registered directory of a language server). I'll check why it doesn't happen when I have enough time.

Thank you for your report!

P.S: if you are interested in fixing the issue yourself, please take a look at activation function creation code. If I were you I would start with debugging the activation function

**Comment by sfavazza** [comment]
I checked once again the `lsp-docker` documentation and your comments. Now I wonder, should the port mapping always be absolute? 

(as you might have noticed, I used `. ` as project folder)

**Comment by sfavazza** [comment]
Hello!

I followed your hint and I observed that the generated `activation-fn` contains code that should have been rendered at generation time. 

Though I forgot to mention that I installed lsp-docker through spacemacs. 

I have to update to the latest version of lsp-docker and verifying whether the issue is solved.

I will try to test the latest version and see whether I can reproduce the reported issue.
